Map/LightOverlay.gd

```diff
@@ -0,0 +1,77 @@
+extends Node2D
+class_name LightOverlay
+
+var color_array: PackedColorArray
+
+var room_container: ColorRect
+var room_ui_array: Array[Array]
+
+var room_polygons_to_draw: Array[Rect2]
+
+func _init(_room_container: ColorRect, _room_ui_array: Array[Array]):
+	room_container = _room_container
+	room_ui_array = _room_ui_array
+
+func _draw():
+	var floor_rect_packed_array: PackedVector2Array
+	var room_circles: Array[PackedVector2Array]
+	var lit_rooms_to_draw: Array[RoomUI]
+	
+	# Account for minimum map placement T____T
+	for floor_index: int in range(room_ui_array.size()):
+		var floor_array: Array = room_ui_array[floor_index]
+		var _first_room_in_array: bool = false
+		
+		for room_index: int in range(floor_array.size()):
+			var room: RoomUI = floor_array[room_index]
+			if room != null:
+				var new_room_texture_rect: TextureRect = Helpers.get_first_child_node_of_type(room, TextureRect)
+				var room_size = new_room_texture_rect.get_size()
+				if (room.get_light_level() != Enums.LightLevel.UNLIT):
+					room_circles.append(_calculate_points_for_circle(room))
+					lit_rooms_to_draw.append(room)
+	floor_rect_packed_array.append(Vector2(0, 0))
+	floor_rect_packed_array.append(Vector2(0, room_container.get_size().y))
+	floor_rect_packed_array.append(Vector2(room_container.get_size().x, room_container.get_size().y))
+	floor_rect_packed_array.append(Vector2(room_container.get_size().x, 0))
+	
+	if room_circles.size() > 0:
+		var current_lit_room_polygon: PackedVector2Array = room_circles[0]
+		for room_circle_index in range(room_circles.size()):
```

Comment:
if I understand this loop correctly, you are basically adding all the lit areas polygons between 1 and the last together ?
maybe you could use Array.slice() to do that ?

Reply:
That is the end product, but to achieve this we merge them individually and add it to the overall final polygon.

First pass through the array we merge the first circle and the second circle together:
![image](https://github.com/Saplings-Projects/1M_sub/assets/5923102/9fe5ae46-b8bf-4c58-ba5b-ab40e707ba25)

This is now one polygon and I save this as the current polygon for all the lit rooms.
Second pass and subsequent passes we merge the next circle to the current polygon then save it as the new current polygon.

In terms of .slice(), I believe it should be something similarly achievable, but slice returns back another Array so there would have to be some extra querying from that new array to get the circles we want to merge. I think what I currently have and just looking at the single array may work best in this case without having to get the logic be too complicated.

Reply:
Alright i see, both should be pretty equivalent, and if it works let's not fix it